### ISSUE TITLE
Meaning of steps in output

### ISSUE THREAD

**Issue by ghost** [viadee/magie#2]
Hi @NoItAll , Thanks for the help yesterday. I was able to generate the output and I went through the readme, however, I was unable to understand why in the output we have explanations under multiple steps for example : step -1, step 0, step 1, step 2, step 3 etc. I understand that for each step we have explanations for both positive class and negative class, however, could not comprehend how to present the final output when we have multiple steps. It will be great if you can explain a bit of result interpretability for easy understanding.

**Comment by NoItAll** [comment]
Hi @adityakumar04, the different steps denote different parts of the pipeline. Look at, for example, the `DirectMemeticMining`-pipeline in the `exemplary_pipeline`-package. In the `addSteps`-method, new steps are added. The `true`-argument for various steps here denotes that the outcome of the step should be persisted. (In the case of the `DirectMemeticMining`-pipeline every step except the last one is to be persisted.) This is done to be able to evaluate the effect of each step. For example: "Does the filter applied possibly eliminate useful rule explanations?" 
A further reason why one would want to keep intermediate rule explanation sets is that some steps have certain properties: For example the `KOptimalRuleExplanationOptimizer` outputs rules which are optimal in the sense that no condition can be deleted or added (with k deletion-/addition-steps) so that the rule explanation is improved on with regards to some metric. If after this `KOptimalRuleExplanationOptimizer` a filter is applied, it might be interesting to see which rules are filtered out (e.g. because they are redundant). 
The last step is the overall output of the pipeline. 
In short: Persisting the result of different steps accounts for the fact that domain knowledge must be applied while evaluating the explanations. 

Please note that persisting explanations might not solely be done via text files. However, it is a feature which I did not have enough time to work on anymore. Currently, it is done using the `TextExplanationPersister`.

I hope you find this answer helpful.

**Comment by ghost** [comment]
Hi @NoItAll Thanks a lot for your explanation. This makes a lot of sense and clears my doubts. Needless to say MAGIE outputs are very intuitive and useful. Thanks for the awesome package.

**Comment by NoItAll** [comment]
@adityakumar04 You are very welcome, I am glad to hear you find it useful!
